src/common/Ulmus.hs

```diff
@@ -46,30 +46,23 @@ initializeApp :: (Show appState, Show action, Read appState, Read action, Typeab
 initializeApp (AppConfig update view updateDisplays initialAppState appPort setup) = do
     setup
     simpleWebApp appPort $ do
-        currentState <- local $ getState initialAppState
-        nextAction <- local (render $ view currentState)
-        currentState' <- local $ getState initialAppState
-        newState <- update nextAction currentState'
-        local (setState newState)
-        renderDisplay initialAppState updateDisplays
-
+        step view
+        loop (step updateDisplays)
+  where
+    step f = do
+      currentState <- local $ getState initialAppState
+      nextAction <- local (render $ f currentState)
+      currentState' <- local $ getState initialAppState
+      newState <- update nextAction currentState'
+      local (setState newState)
+    loop f = f >> loop f
```

Comment:
Ah yeah, here I put the `updateDisplays` step inside an infinite loop, so everytime it's run, `step updateDisplays` waits until `nextAction <- local (render $ f currentState)` actually evaluates to an action, then processes the action, and afterwards enters the same function again, waiting for the next action, it's not an infinite loop that is executed rapidly and eats the cpu, it waits on every action before looping :grin: does that make sense?